[4/n] [c10d] Introduce the multi-tenancy feature in TCPStore

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58331 [4/n] [c10d] Introduce the multi-tenancy feature in TCPStore**
* #58330 [3/n] [c10d] Revise the implementation of TCPStore
* #58329 [2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore
* #58328 [1/n] [c10d] Introduce a new TCPStore constructor

This PR is the final part of a stack that addresses the GitHub issue #41614; it introduces the multi-tenancy feature to the `TCPStore` class allowing two server stores to be instantiated with the same host:port pair.

Differential Revision: [D28453850](https://our.internmc.facebook.com/intern/diff/D28453850/)

cbalioglu committed:
[4/n] [c10d] Introduce the multi-tenancy feature in TCPStore

This PR is the final part of a stack that addresses the GitHub issue #41614; it introduces the multi-tenancy feature to the `TCPStore` class allowing two server stores to be instantiated with the same host:port pair.

Differential Revision: [D28453850](https://our.internmc.facebook.com/intern/diff/D28453850/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[4/n] [c10d] Introduce the multi-tenancy feature in TCPStore"


This PR is the final part of a stack that addresses the GitHub issue #41614; it introduces the multi-tenancy feature to the `TCPStore` class allowing two server stores to be instantiated with the same host:port pair.

Differential Revision: [D28453850](https://our.internmc.facebook.com/intern/diff/D28453850/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -656,12 +657,32 @@ class TCPServer {
+  // If the caller is okay with a multi-tenant store, first check if we already
+  // have a TCPServer running on the specified port.
+  if (opts.multiTenant && opts.port > 0) {
+    auto pos = cachedServers_.find(opts.port);
```
curious, do we need a lock here? Or is multi-thread not a valid use case?

cbalioglu replied:
You are right, we need one. I was too fixated to the single-threaded bootstrap process with `init_process_group()` and `init_rpc()`, and did not consider other cases. Since this code path is potentially shared by more than one store, we need mutual exclusivity. I will fix it in the next rev.

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -671,8 +692,14 @@ std::shared_ptr<TCPServer> TCPServer::start(const TCPStoreOptions& opts) {
+
+  if (opts.multiTenant) {
+    cachedServers_.emplace(port, server);
+  }
```
ditto

mrshenli reviewed (commented):
I haven't fully thought through the concurrency implications. After we allow multi-tenancy, do we need to worry about concurrent calls on `TCPStore::*` APIs?

cbalioglu commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -380,3 +380,23 @@ TEST(TCPStoreTest, testCleanShutdown) {
   clientThread.join();
 }
+
+TEST(TCPStoreTest, testMultiTenantStores) {
```
Yep, will add one in the next rev.

cbalioglu committed:
Update on "[4/n] [c10d] Introduce the multi-tenancy feature in TCPStore"


This PR is the final part of a stack that addresses the GitHub issue #41614; it introduces the multi-tenancy feature to the `TCPStore` class allowing two server stores to be instantiated with the same host:port pair.

Differential Revision: [D28453850](https://our.internmc.facebook.com/intern/diff/D28453850/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[4/n] [c10d] Introduce the multi-tenancy feature in TCPStore"


This PR is the final part of a stack that addresses the GitHub issue #41614; it introduces the multi-tenancy feature to the `TCPStore` class allowing two server stores to be instantiated with the same host:port pair.

Differential Revision: [D28453850](https://our.internmc.facebook.com/intern/diff/D28453850/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[4/n] [c10d] Introduce the multi-tenancy feature in TCPStore"


This PR is the final part of a stack that addresses the GitHub issue #41614; it introduces the multi-tenancy feature to the `TCPStore` class allowing two server stores to be instantiated with the same host:port pair.

Differential Revision: [D28453850](https://our.internmc.facebook.com/intern/diff/D28453850/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[4/n] [c10d] Introduce the multi-tenancy feature in TCPStore"


This PR is the final part of a stack that addresses the GitHub issue #41614; it introduces the multi-tenancy feature to the `TCPStore` class allowing two server stores to be instantiated with the same host:port pair.

Differential Revision: [D28453850](https://our.internmc.facebook.com/intern/diff/D28453850/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/test_c10d_common.py:
```diff
@@ -181,6 +182,45 @@ def test_address_already_in_use(self):
+        # Need to store in an unused variable here to ensure the first
+        # object is not destroyed before the second object is created.
+        store1 = c10d.TCPStore(addr, port, 1, True, multi_tenant=True)  # type: ignore[call-arg] # noqa: F841
+        store2 = c10d.TCPStore(addr, port, 1, True, multi_tenant=True)  # type: ignore[call-arg] # noqa: F841
```
Can we add assertions such as a set by store 1 can be accessed via a get from store2?

cbalioglu replied:
Sure, I have an assertion for that in the C++ test, but I can add it here as well.

rohan-varma commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -769,22 +769,61 @@ class TCPServer {
+    server = startCore();
 
-  auto daemon = std::make_unique<TCPStoreMasterDaemon>(std::move(socket));
+    cachedServers_.emplace(server->port(), server);
```
Do we need this to be a map instead of a single cached server because we want to support `TCPStore::start()` with multiple ports? Curious if we use that feature anywhere or if there are any use cases for it.

cbalioglu replied:
Yes, we need a map because there can be multiple `TCPStore` instances using different ports. For instance:

```python
s1 = TCPStore("localhost", port=1234, multi_tenant=True)
s2 = TCPStore("localhost", port=1234, multi_tenant=True)

s3 = TCPStore("localhost", port=5678, multi_tenant=True)
s4 = TCPStore("localhost", port=5678, multi_tenant=True)
```

Besides being pedantically correct (since we are offering an API and should not make any assumptions about how our types are going to be used), one use case that comes to my mind where this can be useful is TorchElastic. There we can use two multi-tenant `TCPStore` instances, one for the rendezvous itself, and one for the worker. This way we can avoid unnecessarily allocating a second port.

rohan-varma reviewed (approved):
LGTM, thanks! We will finally be able to initialize RPC and c10d without having to open an additional port, and make the user experience better.

cbalioglu committed:
Update on "[4/n] [c10d] Introduce the multi-tenancy feature in TCPStore"


This PR is the final part of a stack that addresses the GitHub issue #41614; it introduces the multi-tenancy feature to the `TCPStore` class allowing two server stores to be instantiated with the same host:port pair.

Differential Revision: [D28453850](https://our.internmc.facebook.com/intern/diff/D28453850/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[4/n] [c10d] Introduce the multi-tenancy feature in TCPStore"


This PR is the final part of a stack that addresses the GitHub issue #41614; it introduces the multi-tenancy feature to the `TCPStore` class allowing two server stores to be instantiated with the same host:port pair.

Differential Revision: [D28453850](https://our.internmc.facebook.com/intern/diff/D28453850/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[4/n] [c10d] Introduce the multi-tenancy feature in TCPStore"


This PR is the final part of a stack that addresses the GitHub issue #41614; it introduces the multi-tenancy feature to the `TCPStore` class allowing two server stores to be instantiated with the same host:port pair.

Differential Revision: [D28453850](https://our.internmc.facebook.com/intern/diff/D28453850/)

[ghstack-poisoned]
